Sensio framework extra4.x

pculka opened the pull request:
fixes https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1768

stof commented on DependencyInjection/Compiler/ConfigurationCheckPass.php:
```diff
@@ -30,7 +30,8 @@ public function process(ContainerBuilder $container)
         if ($container->has('fos_rest.view_response_listener') && isset($container->getParameter('kernel.bundles')['SensioFrameworkExtraBundle'])) {
-            if (!$container->has('sensio_framework_extra.view.listener')) {
+            $config = $container->getExtensionConfig('sensio_framework_extra');
+            if (!(true === $config[0]['view']['annotations'] || $container->has('sensio_framework_extra.view.listener')))  {
```
this is wrong for multiple reasons:

- you check only the first config of the bundle, which is not the final config at all (subsequent configs could enable it, or could re-disable it)
- there is no guarantee at all that `$config[0]['view']['annotations']` exists (the user might be configuring other parts of SensioFrameworkExtraBundle
- this leaves out any processing that the bundle could do on the config (which includes the fact of defining a default value)

pculka replied:
Well in that case the only possible check is to loop through all the configuration and merge it. Default configuration is viewListener enabled in FrameworkExtraBundle...
Meaning the whole view.listener check useless and the sensioframeworkextra bundle dep should be properly written in the documentation?

pculka replied:
Updated the PR to use a xor check against the new TemplateListener which replaces ViewReponseListener in SensioFrameworkExtraBundle  4+ (i.e. it either has new config or old config)

derrabus commented on DependencyInjection/Compiler/ConfigurationCheckPass.php:
```diff
@@ -30,7 +30,7 @@ public function process(ContainerBuilder $container)
 
         if ($container->has('fos_rest.view_response_listener') && isset($container->getParameter('kernel.bundles')['SensioFrameworkExtraBundle'])) {
-            if (!$container->has('sensio_framework_extra.view.listener')) {
+            if (!($container->has('sensio_framework_extra.view.listener') ^ $container->hasDefinition('Sensio\Bundle\FrameworkExtraBundle\EventListener\TemplateListener'))) {
```
Using a bitwise xor operator (`^`) here seems odd. Also, I don't see why the case that both definitions exist should be excluded explicitly. I'd prefer `||` instead.

Also, since we have php 5.5 here, you can safely use `TemplateListener::class` instead of the FQCN string.

merk replied:
If SensioFrameworkExtraBundle is not installed (and isnt required by FOSRestBundle), you cannot use TemplateListener::class because you will end up with the string 'TemplateListener' which is not the same thing.

derrabus replied:
@merk This is wrong. `::class` will always resolve to the FQCN in your `use` statement, no matter if that class exists or not. Also, neither `use` nor `::class` will trigger autoloading. So it is indeed perfectly save to use it here.

https://3v4l.org/RkQNN

derrabus commented:
The tests fail because there's obviously a similar issue with `RequestBodyParamConverter`.

pculka commented:
The xor is there as the configuration is mutually exclusive. Either you use the old verion or a new one. There’s nothing between.
I will check the ParamConverter too

pculka commented:
And also will swap out FQCN

derrabus commented:
> The xor is there as the configuration is mutually exclusive. Either you use the old verion or a new one. There’s nothing between.

True, but there's no reason to force this exclusiveness. If an application (or the FrameworkExtraBundle devs) decide to add an alias to the old service ID as a compatibility layer, there's no reason for FOS REST to fail on this.

pculka commented:
Well, I've updated the code and fixed most of the build except Symfony 3.4 and 4.0, which fail due different tests, not those relevant to this PR.
Can somebody please merge this, so we can start using also newer versions of SymfonyFrameworkExtraBundle?

luispabon commented:
I'm very keen on this as well, just coding something up for a 2018 release and naturally we're working against the very latest

pculka commented:
/poke anyone? Can anybody merge/decline this?

xabbuh commented:
Thank you @pculka.
